zk/stages/stage_batches.go

```diff
@@ -136,23 +166,53 @@ func SpawnStageBatches(
 		return fmt.Errorf("save stage progress error: %v", err)
 	}
 
+	//// BISECT ////
+	if cfg.zkCfg.DebugLimit > 0 && stageProgressBlockNo > cfg.zkCfg.DebugLimit {
+		return nil
+	}
+
 	// get batch for batches progress
 	stageProgressBatchNo, err := hermezDb.GetBatchNoByL2Block(stageProgressBlockNo)
 	if err != nil {
 		return fmt.Errorf("get batch no by l2 block error: %v", err)
 	}
 
-	//// BISECT ////
-	if cfg.zkCfg.DebugLimit > 0 && stageProgressBlockNo > cfg.zkCfg.DebugLimit {
-		return nil
-	}
-
 	highestVerifiedBatch, err := stages.GetStageProgress(tx, stages.L1VerificationsBatchNo)
 	if err != nil {
 		return errors.New("could not retrieve l1 verifications batch no progress")
 	}
 
 	startSyncTime := time.Now()
+
+	latestForkId, err := stages.GetStageProgress(tx, stages.ForkId)
+	if err != nil {
+		return err
+	}
+
+	var dsQueryClient DatastreamClient
+	if cfg.dsQueryClientCreator != nil {
+		dsQueryClient, err = cfg.dsQueryClientCreator(ctx, cfg.zkCfg, latestForkId)
+	} else {
+		dsQueryClient, err = newStreamClient(ctx, cfg.zkCfg, latestForkId)
+	}
+	if err != nil {
+		log.Warn(fmt.Sprintf("[%s] Error when starting datastream client. Error: %s", logPrefix, err))
+		return err
+	}
+	defer dsQueryClient.Stop()
+
+	highestDSL2Block, err := dsQueryClient.GetLatestL2Block()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve the latest datastream l2 block: %w", err)
+	}
```

Comment:
Let's move this to a function.

Reply:
Do you like this approach https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62?

I didn't want to encapsulate `GetLatestL2Block` as well to the function, because the `dsQueryClient` instance is used for other places as well:
- finding the common ancestor
https://github.com/0xPolygonHermez/cdk-erigon/blob/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62/zk/stages/stage_batches.go#L363
- querying the DS for the parent block hash
https://github.com/0xPolygonHermez/cdk-erigon/blob/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62/zk/stages/stage_batches.go#L398

So a separate function would just create and start the DS client.

Reply:
The whole `SpawnStageBatches` function still looks very complex and does too much than it should. If `dsQueryClient` is reused in other places, maybe pass it as a parameter to the functions that will be refactored out, or create a new ds client in those functions.

Reply:
> The whole SpawnStageBatches function still looks very complex and does too much than it should. 

I could not agree more with this statement.

> If dsQueryClient is reused in other places, maybe pass it as a parameter to the functions that will be refactored out, or create a new ds client in those functions.

I'm afraid that it would not help to make this function more manageable even with such a change. I can refactor it entirely into some smaller functions, that are easier for understanding. Though is it wise to do this through this PR (it would make the diff even bigger). WDYT?

Reply:
Agreed to refactor the entire `SpawnStageBatches` through a separate PR.